### PR TITLE
PMM-3594 Send real data to pmm-managed.

### PR DESCRIPTION
**Pull request by AlekSi**

### CODE REVIEW

**Comment by AlekSi** [comment]
@DexterHD Please focus on Go code and overall architecture, and tests.
@askomorokhov Please focus on numbers correctness and tests.

File: Makefile [hunk 1]
```diff
@@ -32,6 +32,11 @@ init:                           ## Installs tools to $GOPATH/bin (which is expec
 gen:                            ## Generate files.
 	go generate ./...
 
+gen-init:
```
Comment (BupycHuk) [inline Makefile]: what does this command do?

File: agents/builtin/mysql/mysql.go [hunk 2]
```diff
@@ -93,78 +114,179 @@ func (m *MySQL) Run(ctx context.Context) {
...
+		}
+
+		buckets[i] = b
 	}
 
-	for _, str := range structs {
-		ess := str.(*eventsStatementsSummaryByDigest)
+	startS := uint32(periodStart.Unix())
+	lengthS := uint32(periodLength.Seconds())
+	for i, b := range buckets {
```
Comment (BupycHuk) [inline agents/builtin/mysql/mysql.go]: why do we use two loops instead of one?

File: agents/builtin/mysql/mysql.go [hunk 3]
```diff
@@ -93,78 +114,179 @@ func (m *MySQL) Run(ctx context.Context) {
...
-		if ess.Digest == nil || ess.DigestText == nil {
-			m.l.Debugf("Skipping %s.", ess)
+	for digest, currentESS := range current {
+		prevESS := prev[digest]
+		if prevESS == nil {
+			prevESS = new(eventsStatementsSummaryByDigest)
+		}
+		count := float32(currentESS.CountStar - prevESS.CountStar)
+		switch {
+		case count == 0:
```
Comment (BupycHuk) [inline agents/builtin/mysql/mysql.go]: is it possible to have the same CountStar value in prevESS and currentESS after Truncate?

File: Makefile (hunk 1 above)
Comment (AlekSi) [reply Makefile]: It generates initial model files (like models.go) that should be edited by hands.
We use only a few tables from performance schema, but it is likely that we will use more in the near future, that's why I kept it for now.

File: agents/builtin/mysql/mysql.go (hunk 2 above)
Comment (AlekSi) [reply agents/builtin/mysql/mysql.go]: Good point, that's a missed refactoring opportunity after functions were merged.

File: agents/builtin/mysql/mysql.go (hunk 3 above)
Comment (AlekSi) [reply agents/builtin/mysql/mysql.go]: Yes, but we can't differentiate those two situations.

File: Makefile (hunk 1 above)
Comment (BupycHuk) [reply Makefile]: Could you add a comment to makefile? Like `## Generate initial model files`

Comment (AlekSi) [reply Makefile]: That comment will make it a public target – visible in `make` output. That target is intended to be private.

File: agents/builtin/mysql/mysql.go [hunk 4]
```diff
@@ -93,78 +114,179 @@ func (m *MySQL) Run(ctx context.Context) {
...
+			{float32(currentESS.SumSortMergePasses - prevESS.SumSortMergePasses), &mb.MMergePassesSum, &mb.MMergePassesCnt},
+			{float32(currentESS.SumSortRange - prevESS.SumSortRange), &mb.MSortRangeSum, &mb.MSortRangeCnt},
+			{float32(currentESS.SumSortRows - prevESS.SumSortRows), &mb.MSortRowsSum, &mb.MSortRowsCnt},
+			{float32(currentESS.SumSortScan - prevESS.SumSortScan), &mb.MSortScanSum, &mb.MSortScanCnt},
+
+			{float32(currentESS.SumNoIndexUsed - prevESS.SumNoIndexUsed), &mb.MNoIndexUsedSum, &mb.MNoIndexUsedCnt},
+			{float32(currentESS.SumNoGoodIndexUsed - prevESS.SumNoGoodIndexUsed), &mb.MNoGoodIndexUsedSum, &mb.MNoGoodIndexUsedCnt},
+		} {
+			if p.value != 0 {
+				*p.sum = p.value
```
Comment (askomorokhov) [inline agents/builtin/mysql/mysql.go]: IMHO, `p` should have more descriptive name here, to easier catch it purposes.

**Comment by DexterHD** [comment]
@AlekSi Is it all about QAN? I think it is.
However, there are only 4 matches of this word and all these words are mentioned only in comments. 
It's weird for me as nothing says what this code is all about (I mean `agents/builtin/mysql`)? Maybe we should use more clean naming? IMHO packages and other stuff names should represent a behavior, not implementation. Maybe it should be `QAN` in package name or in `ServiceName`, I don't know.

File: agents/builtin/mysql/mysql.go [hunk 5]
```diff
@@ -93,78 +125,184 @@ func (m *MySQL) Run(ctx context.Context) {
...
+			// convert picoseconds to seconds
+			{float32(currentESS.SumTimerWait-prevESS.SumTimerWait) / 1e12, &mb.MQueryTimeSum, &mb.MQueryTimeCnt},
+			{float32(currentESS.SumLockTime-prevESS.SumLockTime) / 1e12, &mb.MLockTimeSum, &mb.MLockTimeCnt},
+
+			{float32(currentESS.SumRowsAffected - prevESS.SumRowsAffected), &mb.MRowsAffectedSum, &mb.MRowsAffectedCnt},
+			{float32(currentESS.SumRowsSent - prevESS.SumRowsSent), &mb.MRowsSentSum, &mb.MRowsSentCnt},
+			{float32(currentESS.SumRowsExamined - prevESS.SumRowsExamined), &mb.MRowsExaminedSum, &mb.MRowsExaminedCnt},
+
+			{float32(currentESS.SumCreatedTmpDiskTables - prevESS.SumCreatedTmpDiskTables), &mb.MTmpDiskTablesSum, &mb.MTmpDiskTablesCnt},
+			{float32(currentESS.SumCreatedTmpTables - prevESS.SumCreatedTmpTables), &mb.MQueryTimeSum, &mb.MTmpTableCnt},
```
Comment (askomorokhov) [inline agents/builtin/mysql/mysql.go]: I think here should be:
{float32(currentESS.SumCreatedTmpTables - prevESS.SumCreatedTmpTables), **&mb.MTmpTablesSum,** **&mb.MTmpTablesCnt**},

File: agents/builtin/mysql/mysql.go [hunk 6]
```diff
@@ -49,42 +59,64 @@ type Params struct {
 // Change represents Agent status change _or_ QAN collect request.
 type Change struct {
```
Comment (DexterHD) [inline agents/builtin/mysql/mysql.go]: I don't know, but I think it's bad when one thing in public API represents more than one meaning. These things are different by nature, isn't it? I mean it has different areas of responsibility. `AgentStatus` is about the agent lifecycle but `CollectRequest` is about collecting data.

I mean It's like some function with the name `AddOrJoin(a, b int) int`.

**Comment by AlekSi** [comment]
> However, there are only 4 matches of this word and all these words are mentioned only in comments.

My editors shows 14 occurrences of "qan" in mysql.go 😄. But yeah, I think we should revisit names when we start adding slow log processing.

File: agents/builtin/mysql/mysql.go (hunk 5 above)
Comment (AlekSi) [reply agents/builtin/mysql/mysql.go]: Good catch! Fixed.

File: agents/builtin/mysql/mysql.go (hunk 4 above)
Comment (AlekSi) [reply agents/builtin/mysql/mysql.go]: `p` is a local loop variable, it has to be short. And I can't think of a better name.

File: agents/builtin/mysql/mysql.go (hunk 6 above)
Comment (AlekSi) [reply agents/builtin/mysql/mysql.go]: Indeed, there were two channels. But they were harder to use and much easier to misuse. Supervisor reads changes channel until it is closed. If there are two of them, and one is closed – should we wait for the second to close, or not? Should we drain the data channel before removing that Agent? etc etc. Supervisor code become much more complex.
Again, we could revisit it once we have more other Agents.

**Pull request merged**